### ISSUE TITLE
Emit warnings if undefined function is called on a required module

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/elixir#4540]
Both Elixir and Erlang do not expect the function you are calling to exist during compilation time. However, if a module has been required, we could perform such checks because then, by definition, it is available during compilation time.

The idea of this feature is to check if a function actually exists when compiling code and the function being called belongs to a module that has been required.

**Comment by X4lldux** [comment]
That would surely make our software safer. But specifying all used modules could be tedious, so we could implement a _glob_-like feature:
```Elixir
defmodule SafeModule do
  require *, except: [ModuleA, ModuleB, ModuleC]
  ...
end
```
which would mean, that all "called" modules need to be present at compile time. With a `:except` keyword to specify optional modules.

But I'd rather lean towards requiring all modules, which functions were called, to be present at compile time and give a way to define a list of optional modules. I think this way would cover a lot more use-cases:
```Elixir
defmodule SafeModule do
  @optional_modules [ModuleA, ModuleB]
  @optional_modules ModuleC

  ...
end
```
(this is a potentially breaking change)

**Comment by josevalim** [comment]
@X4lldux that feels to be another separate discussion: how to make it easy to require multiple modules?

I am quite sure we can never make something like * with `:except` work because the only way to know all modules is by compiling all of them. And then if you define require * in two different modules, how to solve that? I believe we should move this discussion elsewhere, possibly the elixir-lang-core mailing list, to keep this one focused on warnings on top of the existing require feature. What do you think?

**Comment by michalmuskala** [comment]
I think such a feature of checking modules in the entire project could be done much easier as a post processing step to the compilation. I'm thinking about something similar to what xref does, but make a convenient mix task of it and promote adding it to the test alias so it runs after (or even before) the tests.

Using `require` for this has also a significant drawback - adding compile-time dependency between modules will most probably significantly decrease compilation speed by limiting the concurrency of the compiler (because you need to compile your dependencies strictly before the particular module is compiled).

We can do a lot more with a separate analysis step - for instance analyzing undefined or unused modules and functions (similar to recently proposed `defmodulep` but based on `@moduledoc false` and `@doc false`). I also think this will be a more extensible solution that will enable adding additional checks later on (I could see a linter integrating into the post-compilation check).

**Comment by josevalim** [comment]
> Using require for this has also a significant drawback 

@michalmuskala that's only if you are using this feature for the purpose of getting warnings.

Maybe another way to put it is: would be worth to emit those warnings given the module was already required? It is considerably cheaper than a precompile check.

**Comment by michalmuskala** [comment]
I'm just afraid that people will start adding additional requires just to get those checks (because they are useful), and they will get slow compilation even not realizing why that is.

**Comment by josevalim** [comment]
I am not sure if it would increase compilation time by any noticiable measure. If one file halts, we will always put another one in the compilation pipeline.

**Comment by michalmuskala** [comment]
The problem is that with something like `require *` you put every file in the compilation pipeline. With single modules I agree this won't be an issue.

I think using a struct from a module also adds compile time dependency - should we do the detection for those modules as well?

Another issue I see is that it might be confusing why you sometimes get those warnings and sometimes you don't.

And to be clear, I'm not against the idea - just trying to be a devil's advocate.

**Comment by josevalim** [comment]
> The problem is that with something like require * you put every file in the compilation pipeline. 

Yes, that's why I asked to move the discussion about `require *` elsewhere. I don't think it could actually work.

> Another issue I see is that it might be confusing why you sometimes get those warnings and sometimes you don't.

The struct is a good point but I think this comment may be enough reason for keeping it only for required modules.

> And to be clear, I'm not against the idea - just trying to be a devil's advocate.

Thank you :)

**Comment by X4lldux** [comment]
After spending some time playing with xref, exref and xref_runner, I would side with @michallepicki. It will be confusing why sometimes compiler warns us and sometimes not. And it would be a lacking duplication of a tool that already exists - xref (it'd be better to integrate it to mix, like eprof was, or maybe even add it to the list of default compilers to be run always after `mix compile`).

**Comment by josevalim** [comment]
@X4lldux well, the issue with integrating with xref is that if it is an optional step, it won't run, and if it is always run, it will slow compilation down. So I like `xref` for its feature set but it is most likely that we would have our own tiny implementation, integrated into Mix, if we ever add something of sorts.

The advantage of relying on require is that the feedback would be immediate and not depend on Mix.

**Comment by X4lldux** [comment]
@josevalim the main issue in my worries, is to not give people the false sense of security. If those warning would be only generated for `require`d modules, we might easily lose track of what module was or wasn't required. And fall in to illusion we didn't make a bug.

If a `xref` detection capabilities would be written into compiler, such it would be able to detect bugs in the entire scope of compiled project, and not just in a subset of it - than I'm all thumbs up!

As for slowing compilation. I've measured it on current HEAD of `ecto` and the only change to `ecto` I've made, was to add my version of `exref` (which should be a little slower because it uses `xref_runner` instead of pure `xref`). I've made two passes of tests with and without `exref` added to compilation process (via `compilers: Mix.compilers ++ [:exref],` in mix.exs). One pass was to always start without `ecto` compiled before, the other was with `ecto` already compiled.
Results are here: https://gist.github.com/X4lldux/976dcc0c85c4e80e60c9e85b4e63ba85
The overall overhead is 1.4s. For such a big project like `ecto`, 1.4s, IMO, is not as big difference that should justify resigning from safety it provides.

Besides, I was proposing to add `xref` to `mix`, on similar condition `eprof` was added (to mark it a good practice) and only as **possibility**, to add it to the default list of compilers :)

**Comment by josevalim** [comment]
1.4s is definitely a lot. We could likely optimize xref to only run if some code was compiled but to me 1.4s is not acceptable for something we would like to make part of Elixir.

**Comment by michalmuskala** [comment]
Yes, that's definitely too much.
I think a separate step might be a good idea. We could generate a task alias like `test.full` or something like that that would run the step along the regular tests.

One way we could speed the process up is to build the dependency graph as we go compiling stuff, but check it only later once all modules are compiled. That would be much more complicated than just calling xref though.

**Comment by X4lldux** [comment]
> We could generate a task alias like test.full or something like that that would run the step along the regular tests.

And run tests and checking of undefined function in parallel, to speed things up, and placing it as a mix task was something I had in mind :+1: 

But nonetheless, we shouldn't be mixing `require` with warning about missed functions. It would be either over-used or over-trusted.
